lib/metrics.js

```diff
@@ -0,0 +1,848 @@
+    if (!this.metric.shouldReport) return;
+
+    if (typeof n === 'object') {
+      meta = n;
+      n = 1;
+    }
+
+    const now = performance.now();
+    this.#window.push({ value: n, time: now });
+
+    let { value } = this;
+    while (this.#window.length && (now - this.#window[0].time) > this.interval) {
+      const cached = this.#window.shift();
+      value -= cached.value;
+    }
+
+    value += n;
+    this.reset(value, meta);
+  }
+}
+
+/**
+ * An increasing or decreasing value.
+ */
+class Counter extends Gauge {
+  /**
+   * Increment the counter. Negative values invert to positive.
+   * @param {number} [n] The amount to increment the counter by. Defaults to 1.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  increment(n = 1, meta) {
+    if (!this.metric.shouldReport) return;
+
+    if (typeof n === 'object') {
+      meta = n;
+      n = 1;
+    }
+
+    this.applyDelta(n, meta);
+  }
+
+  /**
+   * Decrement the counter. Negative values invert to positive.
+   * @param {number} [n] The amount to decrement the counter by. Defaults to 1.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  decrement(n = 1, meta) {
+    if (!this.metric.shouldReport) return;
+
+    if (typeof n === 'object') {
+      meta = n;
+      n = 1;
+    }
+
+    this.applyDelta(-n, meta);
+  }
+}
+
+/**
+ * A floating point number which represents a length of time in milliseconds.
+ */
+class Timer extends Gauge {
+  #meta;
+
+  /**
+   * The start time of the timer.
+   * @property {number} start
+   */
+
+  /**
+   * End time of timer. If undefined, timer is still running.
+   * @property {number|undefined} end
+   */
+
+  /**
+   * Duration of timer in milliseconds. If undefined, timer is still running.
+   * @property {number|undefined} duration
+   */
+
+  /**
+   * Construct a new timer.
+   * @param {Metric} metric The metric to report to.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  constructor(metric, meta) {
+    super(metric);
+    if (meta !== undefined && typeof meta !== 'object') {
+      throw new ERR_INVALID_ARG_TYPE('meta', ['object', 'undefined'], meta);
+    }
+    this.#meta = meta;
+
+    this.start = performance.now();
+    this.end = undefined;
+    this.duration = undefined;
+  }
+
+  /**
+   * Additional metadata to include with the report.
+   * @property {object} meta
+   */
+  get meta() {
+    return mixMeta(this.metric.meta, this.#meta);
+  }
+
+  /**
+   * Stop the timer and report the duration.
+   * @param {object} [meta] Additional metadata to include with the report.
+   * @returns {number} The duration in milliseconds.
+   */
+  stop(meta) {
+    if (this.end !== undefined) return;
+    if (!this.metric.shouldReport) return;
+    this.end = performance.now();
+    this.duration = this.end - this.start;
+    this.reset(this.duration, mixMeta(this.#meta, meta));
+    return this.duration;
+  }
```

Comment:
I'd generally prefer things like `end`, `duration`, etc are prototype getters similar to `meta`.